sys/iec62056/iec62056_21.c

```diff
@@ -0,0 +1,391 @@
+    iec62056_21_dataset_t dataset = { 0 };
+
+    const char *header = NULL;
+    ssize_t res = iec62056_21_get_header(parser, &header);
+
+    if (res < 0) {
+        return res;
+    }
+
+    const char *telegram = header;
+
+    telegram += res;
+    telegram = _skip_space(telegram);
+    len -= telegram - parser->telegram;
+
+    unsigned state = IEC62056_21_STATE_OBIS;
+
+    while (len) {
+        switch (state) {
+        case IEC62056_21_STATE_OBIS:
+        {
+            ssize_t obis_len = _parse_state_obis(parser, &dataset, telegram, len, &skip);
+            if (obis_len < 0) {
+                return obis_len;
+            }
+            len -= obis_len;
+            telegram += obis_len;
+            state = IEC62056_21_STATE_COSEM_OBJECT;
+            break;
+        }
+        case IEC62056_21_STATE_COSEM_OBJECT:
+        {
+            ssize_t val_len = _parse_state_value(parser, &dataset, telegram, len, skip);
+            if (val_len < 0) {
+                return -1;
+            }
+
+            dataset.value_index++;
+            len -= val_len;
+            telegram += val_len;
+            if (len > 0) {
+                const char *next_obj = _skip_space(telegram);
+                int next_state = _next_state(*next_obj);
+                if (next_state < 0) {
+                    return -1;
+                }
+                state = next_state;
+
+                len -= next_obj - telegram;
+                telegram = next_obj;
+            }
+            break;
+        }
+        case IEC62056_21_STATE_LASTLINE:
+            len = 0;
+            break;
+        default:
+            return -1;
+        }
+    }
+    return 0;
+}
+
+bool _register_part_is_numeric(const char *begin, const char *end)
+{
+    while (begin < end) {
+        if (!fmt_is_digit(*begin)) {
+            return false;
+        }
+        begin++;
+    }
+    return true;
+}
+
+bool iec62056_21_dataset_is_register(const iec62056_21_dataset_t *dataset)
+{
+    const char *end_of_register = &dataset->value[dataset->value_length];
+    const char *fractional_part = memchr(dataset->value, '.', dataset->value_length);
+    const char *register_unit = memchr(dataset->value, '*', dataset->value_length);
+
+    if (dataset->value_length == 0) {
+        return false;
+    }
+
+    if (!register_unit) {
+        register_unit = end_of_register;
+    }
+
+    if (!fractional_part) {
+        fractional_part = register_unit;
+    }
+
+    if (!_register_part_is_numeric(dataset->value, fractional_part)) {
+        return false;
+    }
+
+    if (fractional_part == end_of_register) {
+        return true;
+    }
+    fractional_part++;
+
+    return _register_part_is_numeric(fractional_part, register_unit);
+}
+
+int iec62056_21_dataset_parse_register_to_decfrac(const iec62056_21_dataset_t *dataset,
+                                                  int64_t *mantissa, int64_t *scaling)
+{
+    /* parse a string with format '001234.567*kWh' to components */
```

Comment:
This could be made a bit more reusable by moving the number parsing part (`01234.567`) to the `fmt` module - or just by using `sscanf()` for the number